test/middleware/routing_middleware_test.dart

```diff
@@ -0,0 +1,423 @@
+  @override
+  dynamic noSuchMethod(final Invocation invocation) {
+    throw UnimplementedError(invocation.memberName.toString());
+  }
+
+  FakeRequest(final String path,
+      {this.method = relic_req_method.RequestMethod.get})
+      : url = Uri.parse('http://localhost$path');
+}
+
+class FakeResponse implements Response {
+  @override
+  final int statusCode;
+  final Map<Symbol, String>? parameters;
+
+  FakeResponse(this.statusCode, {this.parameters});
+
+  @override
+  dynamic noSuchMethod(final Invocation invocation) {
+    throw UnimplementedError(invocation.memberName.toString());
+  }
+}
+
+void main() {
+  group('RoutingMiddleware', () {
+    late Router<Handler> router;
+    late RoutingMiddleware middleware;
+
+    setUp(() {
+      router = Router<Handler>();
+      middleware = RoutingMiddleware(router);
+    });
+
+    group('Parameter Propagation', () {
+      test(
+          'Given a router with a parameterized route and RoutingMiddleware, '
+          'When a request matches the parameterized route, '
+          'Then the handler receives correct path parameters', () async {
+        Map<Symbol, String>? capturedParams;
+        Future<ResponseContext> testHandler(final RequestContext ctx) async {
+          capturedParams = ctx.pathParameters;
+          return (ctx as RespondableContext)
+              .withResponse(FakeResponse(200, parameters: capturedParams));
+        }
+
+        router.add(Method.get, '/users/:id', testHandler);
+
+        final initialCtx = FakeRequest('/users/123').toContext(Object());
+        final resultingCtx =
+            await middleware.meddle((final RequestContext ctx) async {
+          return (ctx as RespondableContext).withResponse(FakeResponse(404));
+        })(initialCtx);
+
+        expect(capturedParams, isNotNull);
+        expect(capturedParams, equals({#id: '123'}));
+        expect(resultingCtx, isA<ResponseContext>());
+        final response =
+            (resultingCtx as ResponseContext).response as FakeResponse;
+        expect(response.statusCode, equals(200));
+        expect(response.parameters, equals({#id: '123'}));
+      });
+
+      test(
+          'Given a router with a non-parameterized route and RoutingMiddleware, '
+          'When a request matches the non-parameterized route, '
+          'Then the handler receives empty path parameters', () async {
+        Map<Symbol, String>? capturedParams;
+        Future<ResponseContext> testHandler(final RequestContext ctx) async {
+          capturedParams = ctx.pathParameters;
+          return (ctx as RespondableContext)
+              .withResponse(FakeResponse(200, parameters: capturedParams));
+        }
+
+        router.add(Method.get, '/users', testHandler);
+
+        final initialCtx = FakeRequest('/users').toContext(Object());
+        final resultingCtx =
+            await middleware.meddle((final RequestContext ctx) async {
+          return (ctx as RespondableContext).withResponse(FakeResponse(404));
```

Comment:
This reads like it should override the response.


But I guess I am mistaken, and this is just another fallback handler?


I blame `meddle`, which could just mean anything from my point of view 😅🙈

Reply:
I'll refactor to use `Pipeline` and `routeWith` for clarity.

Reply:
Done

Reply:
Thanks!